src/main/java/org/testeditor/web/xtext/index/resources/GlobalScopeResource.xtend

```diff
@@ -0,0 +1,186 @@
+import org.eclipse.emf.ecore.resource.Resource
+import org.eclipse.emf.ecore.resource.ResourceSet
+import org.eclipse.emf.ecore.resource.impl.ResourceSetImpl
+import org.eclipse.xtext.resource.impl.ResourceSetBasedResourceDescriptions
+import org.eclipse.xtext.scoping.IGlobalScopeProvider
+import org.eclipse.xtext.util.StringInputStream
+import org.slf4j.LoggerFactory
+
+@Path("/xtext/index/global-scope")
+@Produces(MediaType.APPLICATION_JSON)
+class GlobalScopeResource implements IGlobalScopeResource {
+
+	protected static val logger = LoggerFactory.getLogger(GlobalScopeResource)
+
+	val IGlobalScopeProvider globalScopeProvider
+	val ResourceSetBasedResourceDescriptions index
+
+	new(IGlobalScopeProvider globalScopeProvider, ResourceSetBasedResourceDescriptions index) {
+		this.globalScopeProvider = globalScopeProvider
+		this.index = index
+	}
+
+	@POST // to allow context content to be passed as payload (instead of using query parameters)
+	@Consumes("text/plain")
+	@Produces("application/json")
+	override Response getScope(String context, @QueryParam("contentType") String contentType,
+		@QueryParam("contextURI") String contextURI, @QueryParam("reference") String eReferenceURIString) {
+		try {
+			val eReference = createEReference(eReferenceURIString)
+			val resource = createContextResource(context, contextURI, contentType)
+
+			logger.debug("Delegating to global scope provider='{}'", globalScopeProvider.class.simpleName)
+			val scope = globalScopeProvider.getScope(resource, eReference, null).allElements
+			logger.debug("Global scope provider returned the following elements='{}'", scope.map[name].join(','))
+
+			return Response.ok(scope.toList).build
+
+		} catch (GlobalScopeResourceException e) {
+			return Response.serverError.entity(e).build
+		}
+	}
+
+	private def createContextResource(String context, String contextURI, String contentType) {
+		logger.debug("Trying to retrieve or create context resource type='{}', URI='{}'",  contentType, contextURI)
+		val resourceSet = new ResourceSetImpl // using empty resource set, since context will be used for this request only and is not part of the index
+		val resource = getOrCreateResource(resourceSet, contextURI, contentType)
+		if (!context.nullOrEmpty) {
+			loadResource(resource, context)
+		}
+		return resource
+	}
+
+	private def getOrCreateResource(ResourceSet resourceSet, String contextURI, String contentType) {
+		try {
+			val uri = URI.createURI(contextURI)
+			val resource = resourceSet.getResource(uri, false) ?: resourceSet.createResource(uri, contentType)
+
+			if (resource !== null) {
+				return resource
+			} else {
+				logger.error('Failure to create get or create resource.')
+				// exception is thrown at the end of the method
+			}
+		} catch (RuntimeException e) {
+			logger.error('Exception during creation of resource.', e)
+		}
+		throw new ResourceCreationException('''Failed to create resource for URI '«contextURI»' of type '«contentType»'.''')
```

Comment:
This looks clumsy.  At first the method does two things. Getting or creating resources. Then the default behaviour is a thrown exception. Default is the case when the method ends (for me).